File: android/src/main/java/com/mcodex/nitroarchive/domain/TaskStateMachine.kt
package com.mcodex.nitroarchive.domain

enum class TaskState {
    IDLE, RUNNING, CANCELLING, SUCCEEDED, FAILED, CANCELLED;

    val isTerminal: Boolean
        get() = this == SUCCEEDED || this == FAILED || this == CANCELLED
}

class TaskStateMachine {
    private var _state: TaskState = TaskState.IDLE
    private val lock = Any()

    val state: TaskState
        get() = synchronized(lock) { _state }

    fun start() {
        synchronized(lock) {
            check(_state == TaskState.IDLE) { "Task is not idle (state=${_state.name})" }
            _state = TaskState.RUNNING
        }
    }

    fun cancel(): Boolean {
        synchronized(lock) {
            return when (_state) {
                TaskState.IDLE -> {
                    _state = TaskState.CANCELLED
                    true
                }
                TaskState.RUNNING -> {
                    _state = TaskState.CANCELLING
                    true
                }
                else -> false
            }
        }
    }

    fun succeed() {
        synchronized(lock) {
            require(_state == TaskState.RUNNING || _state == TaskState.CANCELLING) {
                "Task cannot succeed from state=${_state.name}"
            }
            _state = TaskState.SUCCEEDED
        }
    }

    fun fail() {
        synchronized(lock) {
            _state = TaskState.FAILED
        }
    }

    fun markCancelled() {
        synchronized(lock) {
            require(_state == TaskState.CANCELLING) {
                "Task cannot be cancelled from state=${_state.name}"
            }
            _state = TaskState.CANCELLED
        }
    }
}

File: ios/Domain/TaskStateMachine.swift
import Foundation

final class TaskStateMachine: @unchecked Sendable {
  enum State: String {
    case idle, running, cancelling, succeeded, failed, cancelled
    var isTerminal: Bool {
      switch self {
      case .succeeded, .failed, .cancelled: return true
      case .idle, .running, .cancelling: return false
      }
    }
  }

  private let lock = NSLock()
  private var _state: State = .idle

  var state: State {
    lock.lock()
    defer { lock.unlock() }
    return _state
  }

  func start() throws {
    lock.lock()
    defer { lock.unlock() }
    guard _state == .idle else { throw ArchiveDomainError.invalidState("Task is not idle (state=\(_state.rawValue))") }
    _state = .running
  }

  func cancel() -> Bool {
    lock.lock()
    defer { lock.unlock() }
    switch _state {
    case .idle:
      _state = .cancelled
      return true
    case .running:
      _state = .cancelling
      return true
    case .cancelling, .succeeded, .failed, .cancelled:
      return false
    }
  }

  func succeed() throws {
    lock.lock()
    defer { lock.unlock() }
    guard _state == .running || _state == .cancelling else { throw ArchiveDomainError.invalidState("Task cannot succeed from state=\(_state.rawValue)") }
    _state = .succeeded
  }

  func fail() {
    lock.lock()
    defer { lock.unlock() }
    guard _state == .idle || _state == .running || _state == .cancelling else { return }
    _state = .failed
  }
}
